### PR TITLE
[Integration][Github] - Add Labels Selector to GitHub Issue Kind

### PR DESCRIPTION
**Pull request by shariff-6**
### **User description**
# Description

## What

Adds a new `labels` field to the `GithubIssueSelector` class that allows filtering GitHub issues based on labels during both webhook processing and resync operations. Issues must have **ALL** of the specified labels (AND logic) to be ingested.

## Why

This enhancement provides users with more granular control over which GitHub issues are ingested into Port. Previously, users could only filter issues by state (open/closed/all) and repository query, but couldn't filter based on specific labels. This is particularly useful for:

- Ingesting only issues with certain labels like "bug", "enhancement", or "priority:high"
- Creating focused dashboards for specific issue categories
- Reducing noise by filtering out issues that don't match organizational workflows
- Supporting label-based triage and categorization processes

### Configuration Changes
- Added `labels: Optional[str]` field to `GithubIssueSelector` with description and example usage
- Updated `ListIssueOptions` to include the optional `labels` parameter

### Filtering Logic Implementation
- Created `issue_matches_labels()` utility function in `github/helpers/utils.py` that:
  - Accepts issue labels list and comma-separated required labels string
  - Uses AND logic (all required labels must be present)
  - Handles empty/null inputs gracefully

### Webhook Processing
- Modified `IssueWebhookProcessor.handle_event()` to check labels filter before processing
- Added early return with empty results when labels don't match
- Integrated filtering check before existing state-based logic

### Resync Integration  
- Updated `main.py` resync_issues function to pass `labels` parameter to issue listing operations

### Testing
- Added unit tests in `test_issue_webhook_processor.py` covering:
  - Exact label matches (all required labels present)
  - Partial matches (missing some labels)
  - No labels specified (default behavior)
  - Edge cases with empty strings

## Type of change

Please leave one option from the following and delete the rest:

- [x] New feature (non-breaking change which adds functionality)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Integration testing checklist

- [x] Integration able to create all default resources from scratch
- [x] Completed a full resync from a freshly installed integration and it completed successfully
- [x] Resync able to create entities
- [x] Resync able to update entities
- [x] Resync able to detect and delete entities
- [x] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Enhancement

___

### **Description**
- Add `labels` field to `GithubIssueSelector` for filtering issues by labels

- Implement AND logic requiring all specified labels to be present

- Integrate label filtering into webhook processing and resync operations

- Add comprehensive unit tests for label matching functionality

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["GithubIssueSelector"] -->|"labels field"| B["issue_matches_labels()"]
  B -->|"AND logic"| C["Webhook Processor"]
  B -->|"AND logic"| D["Resync Operations"]
  C -->|"filter issues"| E["WebhookEventRawResults"]
  D -->|"filter issues"| F["ListIssueOptions"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>integration.py</strong><dd><code>Add labels field to issue selector configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/github/integration.py

<ul><li>Added <code>labels</code> optional field to <code>GithubIssueSelector</code> class<br> <li> Field accepts comma-separated label names with AND logic requirement<br> <li> Includes description and example usage in field documentation</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2513/files#diff-352618668ce736511a5de564f4e299286cc93f0445a9f0300368ba6c93452eef">+4/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>options.py</strong><dd><code>Add labels parameter to list options</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/github/github/core/options.py

<ul><li>Added <code>labels</code> parameter to <code>ListIssueOptions</code> class<br> <li> Marked as <code>NotRequired</code> and <code>Optional</code> for backward compatibility</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2513/files#diff-c47b92fb4fc063ecebd6e211ac6c6b269ad22b8e2464fe37ccd838cec4e55e79">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Implement label matching utility function</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/github/github/helpers/utils.py

<ul><li>Created <code>issue_matches_labels()</code> utility function for label filtering<br> <li> Implements AND logic requiring all specified labels to be present<br> <li> Handles edge cases with empty/null inputs gracefully<br> <li> Performs case-insensitive label name comparison</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2513/files#diff-634332229177cd51b278c9204458460bf0e47c06b3e3d22ad413c21d7876ad1e">+27/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>issue_webhook_processor.py</strong><dd><code>Integrate label filtering into webhook processor</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/github/github/webhook/webhook_processors/issue_webhook_processor.py

<ul><li>Added import for <code>GithubIssueSelector</code> and <code>issue_matches_labels</code> function<br> <li> Implemented <code>_check_labels_filter()</code> method to validate issue labels<br> <li> Integrated label filtering check in <code>handle_event()</code> before processing<br> <li> Returns empty results when issue labels don't match selector criteria</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2513/files#diff-57124569c0935b06b417c84d6f646e8d011a2828561e3e9ea58d28feb65f44d1">+17/-3</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Pass labels parameter to resync operations</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/github/main.py

<ul><li>Updated <code>resync_issues()</code> function to pass <code>labels</code> parameter to <br><code>ListIssueOptions</code><br> <li> Ensures label filtering is applied during resync operations</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2513/files#diff-87e9377eb3998df79e7b005e93973b0b888f8431d81b49504bae4264fc148a0a">+1/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_issue_webhook_processor.py</strong><dd><code>Add unit tests for label filtering functionality</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/github/tests/github/webhook/webhook_processors/test_issue_webhook_processor.py

<ul><li>Added <code>TestIssueFiltering</code> test class with four comprehensive test cases<br> <li> Tests exact label matches with AND logic (all required labels present)<br> <li> Tests partial matches (missing some required labels)<br> <li> Tests default behavior when no labels are specified<br> <li> Tests case-insensitive label name matching</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2513/files#diff-c465318a5d235d048dccb4feefb8ad85133e12f03bd281206edce240002252b0">+65/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: integrations/github/integration.py [hunk 1]
```diff
@@ -121,6 +121,10 @@ class GithubIssueSelector(RepoSearchSelector):
         default="open",
         description="Filter by issue state (open, closed, all)",
     )
+    labels: Optional[str] = Field(
+        default=None,
+        description="Filter issues by labels (comma-separated). Issues must have ALL of the specified labels. Example: 'bug,enhancement'",
+    )
```
Comment (mk-armah) [inline integrations/github/integration.py]: Take as a list of string, 
`"".join(labels)` downstream

File: integrations/github/github/webhook/webhook_processors/issue_webhook_processor.py [hunk 2]
```diff
@@ -67,3 +75,9 @@ async def handle_event(
         return WebhookEventRawResults(
             updated_raw_results=[data_to_upsert], deleted_raw_results=[]
         )
+
+    def _check_labels_filter(
+        self, selector: GithubIssueSelector, issue: dict[str, Any]
+    ) -> bool:
+        """Check if issue labels match selector labels filter."""
+        return issue_matches_labels(issue.get("labels", []), selector.labels)
```
Comment (mk-armah) [inline integrations/github/github/webhook/webhook_processors/issue_webhook_processor.py]: Issue was not taken as optional

Comment (mk-armah) [inline integrations/github/github/webhook/webhook_processors/issue_webhook_processor.py]:
```suggestion
        return issue_matches_labels(issue["labels"], selector.labels)
```

File: integrations/github/github/helpers/utils.py [hunk 3]
```diff
@@ -166,3 +166,30 @@ async def enrich_user_with_primary_email(
...
+    if not required_labels:
+        return True
+
+    required_set = {
+        label.strip().lower() for label in required_labels.split(",") if label.strip()
+    }
+    if not required_set:
+        return True
+
+    issue_label_names = {label.get("name", "").lower() for label in issue_labels}
```
Comment (mk-armah) [inline integrations/github/github/helpers/utils.py]:
```suggestion
    issue_label_names = {label["name"].lower() for label in issue_labels}
```

**Review by mk-armah — CHANGES_REQUESTED** [review]
